src/ggml.c

```diff
@@ -133,10 +136,36 @@ static void ggml_print_backtrace(void) {
     if (GGML_NO_BACKTRACE) {
         return;
     }
-    char attach[32];
-    snprintf(attach, sizeof(attach), "attach %d", getpid());
-    int pid = fork();
-    if (pid == 0) {
+#if defined(__linux__)
+    FILE * f = fopen("/proc/self/status", "r");
+    size_t size = 0;
+    char * line = NULL;
+    ssize_t length = 0;
+    while ((length = getline(&line, &size, f)) > 0) {
+        if (!strncmp(line, "TracerPid:", strlen("TracerPid:")) &&
+            (length != strlen("TracerPid:\t0\n") || line[length - 2] != '0')) {
+            // Already being debugged, and the breakpoint is the later abort()
+            free(line);
+            fclose(f);
+            return;
+        }
+    }
+    free(line);
+    fclose(f);
+    int lock[2] = { -1, -1 };
+    pipe(lock);
```

Comment:
Can you add some comments explaining what is the purpose of this and why it is necessary?

Reply:
The lock ensures that gdb is started only after the tracee enabled `PR_SET_PTRACER`. Without the lock, there would be a race condition risking the previous buggy behavior.